### ISSUE TITLE
Problem with strings being converted to unicode

### ISSUE THREAD

**Issue by polosecki** [nipy/nipype#2038]
### Summary
Regular strings seem to be incorrectly converted into unicode (for the Function utility interface at least). This causes functions to fail. Error seems more general than just the Function utiliy interface. 

### Actual behavior
Strings get converted to unicode. This causes functions that expect regular strings to fail
The code below for instance, produces the following error:

>   File "test.py", line 9, in <module>
>     from nipype.interfaces.utility import Function
>   File "/home/opt/anaconda2/lib/python2.7/site-packages/nipype/__init__.py", line 49, in <module>
>     from .pipeline import Node, MapNode, JoinNode, Workflow
>   File "/home/opt/anaconda2/lib/python2.7/site-packages/nipype/pipeline/__init__.py", line 10, in <module>
>     from .engine import Node, MapNode, JoinNode, Workflow
>   File "/home/opt/anaconda2/lib/python2.7/site-packages/nipype/pipeline/engine/__init__.py", line 12, in <module>
>     from .workflows import Workflow
>   File "/home/opt/anaconda2/lib/python2.7/site-packages/nipype/pipeline/engine/workflows.py", line 22, in <module>
>     standard_library.install_aliases()
>   File "/home/opt/anaconda2/lib/python2.7/site-packages/future/standard_library/__init__.py", line 483, in install_aliases
>     import test
>   File "/home/polo/test.py", line 21, in <module>
>     save_txt.run()
>   File "/home/opt/anaconda2/lib/python2.7/site-packages/nipype/interfaces/base.py", line 1081, in run
>     runtime = self._run_wrapper(runtime)
>   File "/home/opt/anaconda2/lib/python2.7/site-packages/nipype/interfaces/base.py", line 1029, in _run_wrapper
>     runtime = self._run_interface(runtime)
>   File "/home/opt/anaconda2/lib/python2.7/site-packages/nipype/interfaces/utility/wrappers.py", line 194, in _run_interface
>     out = function_handle(**args)
>   File "<string>", line 4, in save_txt
>   File "/home/opt/anaconda2/lib/python2.7/site-packages/numpy/lib/npyio.py", line 1143, in savetxt
>     raise ValueError('invalid fmt: %r' % (fmt,))
> ValueError: invalid fmt: u'%.6f'
> Interface Function failed to run.

 ### Expected behavior
Regular strings should remain regular strings?

### How to replicate the behavior
This is the code that reproduces the error:
```
from nipype.interfaces.utility import Function

def save_txt(out_file):
    import numpy as np
    this_array = np.array(['1.1'])
    np.savetxt(out_file, this_array, fmt='%.6f')
    return out_file

save_txt = Function(function=save_txt, input_names='out_file',
                    output_names='out_file')

save_txt.inputs.out_file = 'test.txt'
save_txt.run()
```
Also the following code seem to reproduce the same or related error:
`print(nipype.get_info())`

### Script/Workflow details
None

### Platform details:

Please paste the output of: `python -c "import nipype; print(nipype.get_info()); print(nipype.__version__)"`

I can't run nipype.get_info() !! I get an InterpolationSyntaxError caused also by convertion of string into unicode again!! My nipype.__version__ is u'0.13.0' (notice the Unicode)

### Execution environment
'2.7.13 |Anaconda custom (64-bit)| (default, Dec 20 2016, 23:09:15)
[GCC 4.4.7 20120313 (Red Hat 4.4.7-1)]'

Choose one
- My python environment outside container  (Not sure what this part means)

**Comment by mgxd** [comment]
@polosecki thanks for reporting - this is a python 2.7 bug that happens when calling an external function  that takes a string input within a `Function` - a unicode string is passed through instead. A quick workaround is to explicitly set these strings via `str()`. However, we should look for a better, longterm fix for this. 

As for the `nipype.getinfo()` error, try upgrading to latest master, as I can't replicate on py2.

We highly recommend python 3, and there is a discussion ( #1983 ) on the future of nipype moving towards only py3.

**Comment by polosecki** [comment]
@mgxd Thanks a lot for the quick workaround, it did work. As a more permanent solution for myself, I will try to port my code to Python 3 for the reasons you mention. Upon trying the latest master the `nipype.getinfo()` failure is corrected. Only the `Function` interface is failing now.

**Comment by MariaICL** [comment]
I have the same issue, but I am a newbie to python, and I don't understand the solution to this completely (not sure where to define the str() exactly). My particular example giving me this error is:

from nipype.interfaces.utility import Function

def _gen_index(in_file):
    import numpy as np
    import nibabel as nb
    import os
    out_file = os.path.abspath('index.txt')
    vols = nb.load(in_file).get_data().shape[-1]
    np.savetxt(out_file, np.ones((vols,), dtype=np.int), fmt="%i", newline=" ")
    return out_file

index = Node(name='index',
               interface=Function(input_names=['in_file'],
                                  output_names=['out_file'],
                                  function=_gen_index))

Where exactly should I use the str() workaround?

Thanks!

**Comment by mgxd** [comment]
@MariaICL 

```
np.savetxt(out_file, np.ones((vols,), dtype=np.int), fmt=str("%i"), newline=str(" "))
```

**Comment by MariaICL** [comment]
Thank you!

**Comment by gauvinalexandre** [comment]
Hi @mgxd ! I just want to tell you I'm experiencing the same issue right now, forcing me to use nipype 0.12.0. I saw that your current PR is subject to discussion due to print function removal. However, is there a way to separate your PR in two in order to fix this issue without wait for a solution for the other? Thanks a lot for the support!

**Comment by mgxd** [comment]
@gauvinalexandre if you'd like to use current nipype with py2 support and don't want to wait for the PR to be merged, you could install my branch with 
```
pip install -U https://github.com/mgxd/nipype/archive/fix/py2functions.zip
```

**Comment by gauvinalexandre** [comment]
Hi! Thanks for your reply! Of course, I can do this, I was simply wondering
about a long term solution since my research team would be interested in
your fix, since we still rely on py2 dependent tools. I'm going to follow
the advancement of your work. Thanks!

**Comment by marieligit** [comment]
Dear @mgxd 

I'm experiencing the same issue and I am a newbie of python (same situation as previously written by MariaICL). I don't know where to add the `str()` fix exactly, and I'd love getting any help.

In my particular case I'm using `MRTrix2TrackVis` by following this example:
`>>> import nipype.interfaces.mrtrix as mrt >>> tck2trk = mrt.MRTrix2TrackVis()`


>>> tck2trk.inputs.in_file = 'dwi_CSD_tracked.tck'
>>> tck2trk.inputs.image_file = 'diffusion.nii'
>>> tck2trk.run()

**Comment by marieligit** [comment]
Dear @mgxd

I'm experiencing the same issue and I am a newbie of python (same situation as previously written by **MariaICL**). I don't know where to add the `str()` fix exactly, and I'd love getting any help.

In my particular case I'm using `MRTrix2TrackVis` from nipype (I guess it uses nipype/nipype/interfaces/mrtrix/convert.py)  by following this example:
 
```
>>> import nipype.interfaces.mrtrix as mrt
>>> tck2trk = mrt.MRTrix2TrackVis()
>>> tck2trk.inputs.in_file = 'dwi_CSD_tracked.tck'
>>> tck2trk.run()   
```

I am getting the following error:
```
...
File "/home/brain/miniconda2/lib/python2.7/genericpath.py", line 26, in exists
    os.stat(path)
TypeError: coercing to Unicode: need string or buffer, _Undefined found
Interface MRTrix2TrackVis failed to run. 

```

my Python version is 2.7.13, my nipype version is 0.12.1.

Any help would be highly appreciated. Thank you in advance!

**Comment by djarecka** [comment]
@mgxd - was it solved in your merged  PR (#2093)

**Comment by mgxd** [comment]
it was reverted - #2150 is the PR that needs review for this.
